### ISSUE TITLE
Failure to launch multiple instances of ethminer on multiple 6gb NVIDIA cards

### ISSUE THREAD

**Issue by tczaugg** [ethereum-mining/ethminer#2097]
**Describe the bug**
There is a bug that prevents launching an instance of ethminer on each of several Nvidia cards. The bug is on line 293 of libethash-cuda/CUDAMiner.cpp where the program calls cudaMemGetInfo without setting the cuda device to i.  So if an instance of ethminer has already been launched on card 0, it will fail to launch on any other card because if doesn't think there is enough free memory, as the deviceDescriptor has been populated with the wrong information (i.e. the free memory for card 0 instead of i).  My cards are GTX 1060s with 6 GB memory.  Once the DAG is created, there is not enough memory on card 0 for another DAG so the check of required memory vs free memory fails.

**To Reproduce**
launch ethminer with --cuda-devices 0, then lauch ethminer with --cuda-devices 1 (or any other number greater than 0)
note that if you lauch on card 1 first you can still launch on card 0 as the deviceDescriptor is always populated with the memory for card 0.

**Expected behavior**
I used to run a separate instance of ethminer for each card.  This way I could tune each card individually and if it crashed I just had to restart that card instead of reboot.  But now the DAG is big enough to expose the bug.

**Solution**
This can be fixed by adding the line
CUDA_SAFE_CALL(cudaSetDevice(i));
before the call to cudaMemGetInfo.

**Comment by invidtiv** [comment]
This has been commented before, search previous issues , this solution is not advised, it is not a bug , use one instance of etherminer for all graphics card, running multiple instance will slow down your overall hashrate, besides launching that error.

**Comment by tczaugg** [comment]
It is a bug.  You're getting the wrong value for the free memory on the card you're creating a descriptor for.  Nor does is slow down your hash rate.  I have implemented this fix and have one board with 13 cards running with 13 Ethminer instances and it is working fine.  I used to do this with all my boards 3 years ago.  Now, however, Ethminer takes up 1.5 GB of resident memory for each instance, so I can only run this way on the system I have with 32 GB.  I used to be able to run this way on boards with 4 GB.  What is taking up so much ram on the CPU?

**Comment by chigui271** [comment]
So, to see if I got this right, @tczaugg ,you are saying that I could do something like this: (with my 6 GPUs)
ethminer -F http://eth1.nanopool.org:8888/myaddresshere --no-precompute --cuda-devices 5
ethminer -F http://eth1.nanopool.org:8888/myaddresshere --no-precompute --cuda-devices 4
ethminer -F http://eth1.nanopool.org:8888/myaddresshere --no-precompute --cuda-devices 3
ethminer -F http://eth1.nanopool.org:8888/myaddresshere --no-precompute --cuda-devices 2
ethminer -F http://eth1.nanopool.org:8888/myaddresshere --no-precompute --cuda-devices 1
ethminer -F http://eth1.nanopool.org:8888/myaddresshere --cuda-devices 0

and I would still be fine?  and no degradation on hash rates?

thanks

**Comment by tczaugg** [comment]
Yes, except without the no-precompute.  The instances don't communicate the DAG and have to compute their own.  But as long as you have enough memory on your CPU it works.  I have a system with 13 cards and 32 GB of memory.  Currently the system has been up for 9 days straight.  All my GPUs are GTX 1060 with 6 GB ram.  I'm getting a hash rate of 290 Mh/s for this system.  I have another system with the same brand of cards but only 4 GB of memory, so I have to run a single instance.  I'm getting a hash rate of 279 Mh/s on that system.  So I think that is pretty good evidence that there is no degradation from running multiple instances of ethminer.
I did modify the code as I suggested so I don't have to launch them in reverse order.  I have a program that monitors each card and restarts it if one fails.  When I run a single instance I usually have to reboot if there is a failure.  I'd much rather restart a single card than all of them.

**Comment by chigui271** [comment]
Thanks.. Great insights. I concurr.
Now, 290h/s for the whole system is about a 22 avg, per GPU right?  Doesn't that seem a bit low?  I was able to get 30-32 (up to 40s) with just one NVIDIA Tesla P100.  Maybe its the extra memory on the Tesla - cause they are pretty similar overall I think.

**Comment by stevebanik** [comment]
@chigui271 I'm using a Tesla T4 and I only get about 13MH/s.  I figured I would get closer to 20-30.  I'm wondering if I need to apply some non-default settings to ethminer.  If anyone has any suggestions, I'd love to hear about them.

**Comment by av3ri** [comment]
> Yes, except without the no-precompute.  The instances don't communicate the DAG and have to compute their own.  But as long as you have enough memory on your CPU it works.  I have a system with 13 cards and 32 GB of memory.  Currently the system has been up for 9 days straight.  All my GPUs are GTX 1060 with 6 GB ram.  I'm getting a hash rate of 290 Mh/s for this system.  I have another system with the same brand of cards but only 4 GB of memory, so I have to run a single instance.  I'm getting a hash rate of 279 Mh/s on that system.  So I think that is pretty good evidence that there is no degradation from running multiple instances of ethminer.
> I did modify the code as I suggested so I don't have to launch them in reverse order.  I have a program that monitors each card and restarts it if one fails.  When I run a single instance I usually have to reboot if there is a failure.  I'd much rather restart a single card than all of them.

what program do you use to monitor each card and restart the instance if one fails? Do you mind sharing your wisdom?

**Comment by tczaugg** [comment]
I wrote my own python program called foreman.  It can also be used to tune
the memory clock rate for each card so as to maximize the hash rate.  Much
easier than tuning by hand.  But without being able to run multiple
instances it doesn't work as well.  Tuning involves pushing the card until
it fails and then backing off to where failures are infrequent.  However,
when running all as one instance, every failure requires a reboot,
affecting all the cards , which makes autotuning less attractive.
